include/kodlab_mjbots_sdk/io_behavior.h

```diff
@@ -0,0 +1,178 @@
+ * @note To include additional LCM behavior inputs in child classes, simply
+ * compose additional `LcmMessageHandler` objects and call
+ * `subscriber_->AddSubscription` for each additional subscription when
+ * initializing the behavior.
+ * @note To include additional LCM  behavior outputs in child classes, simply
+ * compose additional `LcmPublisher` objects and initialize them with the
+ * behavior.
+ * @tparam Robot[optional] `kodlab::RobotBase`-derived robot class
+ * @tparam Input[optional] LCM input message class, default is `VoidLcm`
+ * @tparam Output[optional] LCM output message class, default is `VoidLcm`
+ */
+template<class Robot = kodlab::RobotBase, class Input = VoidLcm,
+    class Output = VoidLcm>
+class IOBehavior : public Behavior<Robot> {
+
+ public:
+
+  /**
+   * @brief Construct an I/O Behavior object
+   * @param robot robot that behavior is executing on
+   * @param subscriber realtime LCM subscriber to monitor for inputs
+   * @param publisher LCM publisher object to publish outputs with
+   * @param subscribe_channel LCM channel to subscribe to
+   * @param name behavior name
+   */
+  IOBehavior(std::shared_ptr<Robot> robot,
+             std::shared_ptr<LcmSubscriber> subscriber,
+             LcmPublisher<Output> publisher,
+             std::string subscribe_channel,
+             std::string name = "")
+      : Behavior<Robot>(robot, std::move(name)),
+        subscriber_(subscriber),
+        publisher_(publisher),
+        publisher_data_(publisher_.get_message_shared_ptr()) {
+    subscriber_->AddSubscription<Input>(subscribe_channel,
+                                        input_handler_);
+  }
+
+  /**
+   * @brief Construct an I/O Behavior object
+   * @param robot robot that behavior is executing on
+   * @param subscriber realtime LCM subscriber to monitor for inputs
+   * @param publish_lcm LCM object used for publishing
+   * @param subscribe_channel LCM channel to subscribe to
+   * @param publish_channel LCM channel to publish on
+   * @param name behavior name
+   */
+  IOBehavior(std::shared_ptr<Robot> robot,
+             std::shared_ptr<LcmSubscriber> subscriber,
+             std::shared_ptr<lcm::LCM> publish_lcm,
+             std::string subscribe_channel,
+             std::string publish_channel,
+             std::string name = "")
+      : IOBehavior(robot,
+                   subscriber,
+                   std::move(LcmPublisher<Output>(publish_lcm, publish_channel)),
+                   subscribe_channel,
+                   name) {}
+
+  /**
+   * Virtual destructor for clean destruction of child class instances
+   */
+  virtual ~IOBehavior() = default;
+
+  /**
+   * @brief Processes input data, if available
+   */
+  void ProcessInput() final {
+    auto input_msg = input_handler_.GetDataIfNew();
+    if (input_msg.has_value()) {
+      ProcessInput(input_msg.value());
+      input_data_ = input_msg.value();
+    }
+  }
```

Comment:
The SFINAE solution here requires that the virtual method in `Behavior` must have the same template, and this is not the desired behavior since the user may specify input type `VoidLcm` in `Behavior`, but still be making use of the `ProcessInput()` method for non-LCM inputs.  

A better solution would be to have two different, private `ProcessInputImpl` methods defined in `IOBehavior`, and select an implementation in `ProcessInput` based on tag dispatching.

For our purposes, I think the additional overhead of the simple approach is negligible relative to tag dispatching, and better maintains code readability.

Reply:
Simple version added in 7a9ebe09f21125317b19bed88d7395002bbd1a41